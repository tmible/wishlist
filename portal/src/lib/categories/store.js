import { initCategories } from '$lib/categories/use-cases/init-categories.js';

/** @typedef {import('./domain.js').Category} Category */

/** @module Адаптер хранилища категорий */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {Category[]}
 */
let value = [];

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * категорий
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => Category[]} get Получение значения хранилища
 * @property {(newValue: Category[]) => void} set Установка значения хранилища
 * @property {(category: Category) => void} add Добавление категории в хранилище
 * @property {(category: Category) => void} update Обновление категории в хранилище
 * @property {(category: Category) => void} delete Удаление категории из хранилища
 */
export const categories = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);

    if (subscriptions.size === 1) {
      initCategories();
    }

    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  add: (category) => {
    value.push(category);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  update: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value[categoryIndex] = category;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  delete: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value.splice(categoryIndex, 1);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
