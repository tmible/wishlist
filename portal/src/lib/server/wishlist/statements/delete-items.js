import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { Database } from '$lib/server/db/injection-tokens.js';
import { DeleteItems } from '../events.js';

/**
 * Создание и выполнение SQL выражений для удаления нескольких элементов списка желаний
 * @function deleteItems
 * @param {number} userid Идентифкатор пользователя — владельца списка
 * @param {number[]} ids Идентификаторы удаляемых элементов списка
 * @returns {{ changes: number }} Количество удалённых элементов списка
 */
const deleteItems = (userid, ids) => {
  if (ids.length <= 0) {
    return { changes: 0 };
  }

  const db = inject(Database);

  const idPlaceholders = ids.map(() => '?').join(', ');
  const statements = [
    `DELETE FROM description_entities WHERE list_item_id IN (${idPlaceholders})`,
    `DELETE FROM participants WHERE list_item_id IN (${idPlaceholders})`,
    `DELETE FROM list WHERE id IN (${idPlaceholders}) AND userid = ?`,
  ];

  for (const statement of statements.slice(0, -1)) {
    db.prepare(statement).run(ids);
  }

  return db.prepare(statements.at(-1)).run(...ids, userid);
};

/**
 * Подписка [выполнения SQL выражений]{@link deleteItems} на соответствующее событие
 * @function initDeleteItemsStatements
 * @returns {void}
 */
export const initDeleteItemsStatements = () => {
  subscribe(DeleteItems, deleteItems);
};
