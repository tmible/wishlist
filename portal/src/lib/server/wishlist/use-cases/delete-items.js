import { inject } from '@tmible/wishlist-common/dependency-injector';
import { emit } from '@tmible/wishlist-common/event-bus';
import {
  LACK_OF_DATA_ERROR_MESSAGE,
} from '$lib/server/constants/lack-of-data-error-message.const.js';
import { RunStatementAuthorized } from '$lib/server/db/events.js';
import { IPCHub } from '$lib/server/ipc-hub/injection-tokens.js';
import { DeleteItems } from '../events.js';

/** @module Сценарий удаления нескольких элементов из списка желаний */

/**
 * Удаление нескольких элементов из списка желаний
 * @function deleteItems
 * @param {number} userid Идентифкатор пользователя — владельца списка
 * @param {number[]} ids Идентификаторы удаляемых элементов списка
 * @returns {void}
 * @throws {Error} Ошибка в случае передачи пустого массива ids
 */
export const deleteItems = (userid, ids) => {
  if (ids.length <= 0) {
    throw new Error(LACK_OF_DATA_ERROR_MESSAGE);
  }
  emit(RunStatementAuthorized, () => emit(DeleteItems, userid, ids), ids.length);
  inject(IPCHub).sendMessage(`update ${userid}`);
};
