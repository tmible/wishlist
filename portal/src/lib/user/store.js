import { DEFAULT_USER } from './default-user.const.js';

/** @typedef {import('./domain.js').User} User */

/** @module Адаптер хранилища пользователя */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {User}
 */
let value = DEFAULT_USER;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * пользователя
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => User} get Получение значения хранилища
 * @property {(newValue: User) => void} set Установка значения хранилища
 */
export const user = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },
  get: () => value,
  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
