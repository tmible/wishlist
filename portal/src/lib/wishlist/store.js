/** @typedef {import('./domain.js').OwnWishlistItem} OwnWishlistItem */
/** @typedef {import('./use-cases/reorder-list.js').ReorderPatch} ReorderPatch */

/** @module Адаптер хранилища списка желаний */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {OwnWishlistItem[] | null}
 */
let value = null;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * Уведомление подписчиков об изменении значения хранилища
 * @function notifySubscribers
 * @returns {void}
 */
const notifySubscribers = () => {
  for (const subscriber of subscriptions.values()) {
    subscriber(value);
  }
};

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * списка желаний
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => OwnWishlistItem[]} get Получение значения хранилища
 * @property {(newValue: OwnWishlistItem[]) => void} set Установка значения хранилища
 * @property {(item: OwnWishlistItem) => void} add Добавление элемента списка в хранилище
 * @property {(item: OwnWishlistItem) => void} update Обновление элемента списка в хранилище
 * @property {(itemIds: OwnWishlistItem['id'][]) => void} delete
 *   Удаление элементов списка из хранилища
 * @property {(patch: ReorderPatch) => void} reorder Переупорядочивание списка в хранилище
 */
export const wishlist = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    notifySubscribers();
  },

  add: (item) => {
    if (!value) {
      value = [];
    }
    value.push(item);
    notifySubscribers();
  },

  update: (item) => {
    const itemIndex = value?.findIndex(({ id }) => id === item.id);
    if (itemIndex === -1) {
      return;
    }
    value[itemIndex] = item;
    notifySubscribers();
  },

  delete: (itemIds) => {
    let isAnyDeleted = false;

    for (const id of itemIds) {
      const itemIndex = value?.findIndex((item) => id === item.id);
      if (itemIndex === -1) {
        continue;
      }
      value.splice(itemIndex, 1);
      isAnyDeleted = true;
    }

    if (!isAnyDeleted) {
      return;
    }

    notifySubscribers();
  },

  reorder: (patch) => {
    if (!value) {
      return;
    }
    for (const item of value) {
      const reorder = patch.find(({ id }) => id === item.id);
      item.order = reorder?.order ?? item.order;
    }
    value.sort((a, b) => a.order - b.order);
    notifySubscribers();
  },
};
