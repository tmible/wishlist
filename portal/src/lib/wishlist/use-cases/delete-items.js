import { inject } from '@tmible/wishlist-common/dependency-injector';
import { NetworkService, Store } from '../injection-tokens.js';

/** @typedef {import('../domain.js').OwnWishlistItem} OwnWishlistItem */

/** @module Сценарий удаления нескольких элементов списка желаний */

/**
 * Запрос на удаление нескольких элементов списка, обновление хранилища
 * @function deleteItems
 * @param {OwnWishlistItem['id'][]} itemIds Удаляемые элементы
 * @returns {Promise<void>}
 * @async
 */
export const deleteItems = async (itemIds) => {
  if (itemIds.length === 0) {
    return;
  }

  const [ , ok ] = await inject(NetworkService).deleteItems(itemIds);

  if (!ok) {
    return;
  }

  inject(Store).delete(itemIds);
};
