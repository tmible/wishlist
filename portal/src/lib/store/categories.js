import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */
/**
 * Категория элементов списка желаний пользователя
 * @typedef {object} Category
 * @property {number} id Идентификатор категории
 * @property {string} name Название категории
 */

/**
 * Svelte хранилище категорий элементов списка желаний пользователя
 * @type {Writable<Category[] | null>}
 */
export const categories = writable(null, (set, update) => {
  let timeout;

  update((value) => {
    if (value === null) {
      timeout = setTimeout(async () => set(await fetch(
        '/api/wishlist/categories',
      ).then(
        (response) => response.json(),
      )));
    }
    return value;
  });

  return () => timeout && clearTimeout(timeout);
});
