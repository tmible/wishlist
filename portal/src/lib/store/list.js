import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */
/** @typedef {import('@tmible/wishlist-common/constants/list-item-state').default} ListItemState */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/**
 * Элемент списка желаний пользователя во варианте отображения для владельца
 * @typedef {object} OwnListItem
 * @property {number} id Идентификатор элемента
 * @property {number} priority Приоритет элемента
 * @property {string} name Название подарка
 * @property {string} description Описание подарка
 * @property {ListItemState} state Состояние подарка
 * @property {Entity[]} descriptionEntities Элементы разметки текста описания подарка
 */

/**
 * Svelte хранилище списка желаний пользователя
 * @type {Writable<OwnListItem[] | null>}
 */
export const list = writable(null);
