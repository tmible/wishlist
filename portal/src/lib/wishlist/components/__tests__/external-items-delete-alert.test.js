// @vitest-environment jsdom
import { cleanup, render, screen } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { writable } from 'svelte/store';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { deleteItems } from '../../use-cases/delete-items.js';
import ExternalItemsDeleteAlert from '../external-items-delete-alert.svelte';

vi.mock(
  '../../store.js',
  () => ({ wishlist: writable([{ isExternal: 0, id: 1 }, { isExternal: 1, id: 2 }]) }),
);
vi.mock('../../use-cases/delete-items.js');

describe('wishlist / components / external items delete alert', () => {
  let modalPortal;

  beforeEach(() => {
    modalPortal = document.createElement('div');
    modalPortal.id = 'modal-portal';
    document.body.append(modalPortal);
  });

  afterEach(() => {
    modalPortal.remove();
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed closed', () => {
    const { baseElement } = render(ExternalItemsDeleteAlert, { open: false });
    expect(baseElement).toMatchSnapshot();
  });

  describe('opened', () => {
    let baseElement;

    beforeEach(() => {
      ({ baseElement } = render(ExternalItemsDeleteAlert, { open: true }));
    });

    it('should be displayed', () => {
      expect(baseElement).toMatchSnapshot();
    });

    it('should close on cancel button click', async () => {
      const user = userEvent.setup();
      await user.click(screen.getByText('Отмена', { selector: 'button' }));
      expect(baseElement).toMatchSnapshot();
    });

    describe('on confirm button click', () => {
      beforeEach(async () => {
        const user = userEvent.setup();
        await user.click(screen.getByText('Удалить', { selector: 'button' }));
      });

      it('should close', () => {
        expect(baseElement).toMatchSnapshot();
      });

      it('should request deletion', () => {
        expect(vi.mocked(deleteItems)).toHaveBeenCalledWith([ 2 ]);
      });
    });
  });
});
