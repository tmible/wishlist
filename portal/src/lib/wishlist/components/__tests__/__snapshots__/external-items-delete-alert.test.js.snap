// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wishlist / components / external items delete alert > opened > on confirm button click > should close 1`] = `
<body
  style="margin-right: 8px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-24"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-state="closed"
      id="bits-22"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`wishlist / components / external items delete alert > opened > should be displayed 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="modal-backdrop"
      data-alert-dialog-overlay=""
      data-state="open"
      id="bits-5"
      style="pointer-events: auto;"
    >
      <!---->
    </div>
    
    
    
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
     
    <div
      aria-describedby="bits-10"
      aria-labelledby="bits-9"
      aria-modal="true"
      class="modal-box flex flex-col justify-center"
      data-alert-dialog-content=""
      data-state="open"
      id="bits-6"
      role="alertdialog"
      style="pointer-events: auto; outline: none;"
      tabindex="-1"
    >
      <div
        class="contents prose"
      >
        <!---->
        <div
          aria-level="2"
          data-alert-dialog-title=""
          data-state="open"
          id="bits-9"
          role="heading"
        >
          Подвердите удаление
          <!---->
        </div>
        
        <!---->
         
        <!---->
        <!---->
        <p
          data-alert-dialog-description=""
          data-state="open"
          id="bits-10"
        >
          Вы уверены, что хотите удалить все сюрпризы из списка?
        </p>
        
        
        <!---->
         
        <div
          class="card-actions"
        >
          <!---->
          <button
            class="btn btn-neutral flex-1"
            data-alert-dialog-cancel=""
            data-state="open"
            id="bits-11"
            tabindex="0"
          >
            Отмена
            <!---->
          </button>
          
          <!---->
           
          <!---->
          <button
            class="btn btn-error flex-1"
            data-alert-dialog-action=""
            data-state="open"
            id="bits-12"
          >
            Удалить
            <!---->
          </button>
          
          <!---->
        </div>
      </div>
      <!---->
    </div>
    
    
    
    
    
    
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-6"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-state="open"
      id="bits-4"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`wishlist / components / external items delete alert > opened > should close on cancel button click 1`] = `
<body
  style=""
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-15"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-state="closed"
      id="bits-13"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`wishlist / components / external items delete alert > should be displayed closed 1`] = `
<body>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-state="closed"
      id="bits-1"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;
